File: automedal/agent/retry.py
"""Transient-error retry helper for provider / advisor calls.

Wraps an async callable with N attempts and backoff delays. Matches
ml-intern's posture (agent_loop.py L118-L136 / L295-L334):

  • Retry only on known-transient patterns: 5xx, 429, timeout, connection reset.
  • Never retry on auth / credits / model-not-found — those are deterministic.
  • Emit one `tool_log`-style event per retry so the TUI shows what's happening.
  • Max 3 attempts with [5, 15, 30]s backoff — ≤50s total before surfacing.
"""

from __future__ import annotations

import asyncio
from typing import Any, Awaitable, Callable, TypeVar

_MAX_ATTEMPTS = 3
_RETRY_DELAYS: tuple[int, ...] = (5, 15, 30)

_TRANSIENT_PATTERNS = (
    "timeout", "timed out",
    "429", "rate limit", "rate_limit",
    "503", "service unavailable",
    "502", "bad gateway",
    "500", "internal server error",
    "504", "gateway timeout",
    "overloaded", "capacity",
    "connection reset", "connection refused", "connection error",
    "eof", "broken pipe",
    "remote end closed",
)


def is_transient_error(exc: BaseException) -> bool:
    """True if the exception matches a known transient network/provider pattern."""
    if isinstance(exc, (asyncio.TimeoutError, TimeoutError)):
        return True
    s = str(exc).lower()
    return any(p in s for p in _TRANSIENT_PATTERNS)


T = TypeVar("T")


async def with_retry(
    call: Callable[[], Awaitable[T]],
    *,
    label: str,
    events: Any = None,
    attempts: int = _MAX_ATTEMPTS,
    delays: tuple[int, ...] = _RETRY_DELAYS,
) -> T:
    """Run `call()`; on transient error, sleep-and-retry up to `attempts` times.

    ``label`` identifies the call in retry events (e.g. 'openai.chat_stream').
    ``events`` is an optional EventSink with a ``tool_log(tool, log)`` method.
    Non-transient exceptions bubble immediately.
    """
    last_exc: BaseException | None = None
    for attempt in range(1, attempts + 1):
        try:
            return await call()
        except Exception as exc:
            last_exc = exc
            if attempt >= attempts or not is_transient_error(exc):
                raise
            delay = delays[min(attempt - 1, len(delays) - 1)]
            if events is not None:
                try:
                    events.tool_log(
                        tool="retry",
                        log=(
                            f"{label}: transient error (attempt {attempt}/{attempts}) "
                            f"— retrying in {delay}s: {type(exc).__name__}: {exc}"
                        ),
                    )
                except Exception:
                    pass
            await asyncio.sleep(delay)
    # Unreachable — either we returned or raised — but keep type-checkers happy.
    assert last_exc is not None
    raise last_exc

File: internal/agent/retry.go
package agent

import (
	"context"
	"errors"
	"fmt"
	"net"
	"strings"
	"time"
)

const (
	defaultRetryAttempts = 3
)

var defaultRetryDelays = []time.Duration{5 * time.Second, 15 * time.Second, 30 * time.Second}

// transient patterns matched against the lowercased exception text — same
// list as automedal/agent/retry.py:_TRANSIENT_PATTERNS.
var transientPatterns = []string{
	"timeout", "timed out",
	"429", "rate limit", "rate_limit",
	"503", "service unavailable",
	"502", "bad gateway",
	"500", "internal server error",
	"504", "gateway timeout",
	"overloaded", "capacity",
	"connection reset", "connection refused", "connection error",
	"eof", "broken pipe",
	"remote end closed",
}

// IsTransientError reports whether err matches a known transient
// network/provider pattern that's worth retrying.
func IsTransientError(err error) bool {
	if err == nil {
		return false
	}
	// Hard-typed transients first.
	if errors.Is(err, context.DeadlineExceeded) {
		return true
	}
	var ne net.Error
	if errors.As(err, &ne) && ne.Timeout() {
		return true
	}
	s := strings.ToLower(err.Error())
	for _, p := range transientPatterns {
		if strings.Contains(s, p) {
			return true
		}
	}
	return false
}

// RetryOpts customises the backoff policy.
type RetryOpts struct {
	Attempts int
	Delays   []time.Duration
	Sink     *EventSink // optional, for retry-attempt notices
	Label    string     // identifies the call site in notices
}

// WithRetry runs `fn` up to opts.Attempts times, sleeping the corresponding
// opts.Delays[i-1] (clamped to last) between attempts. Non-transient errors
// bubble immediately. Honors ctx cancellation.
//
// Mirrors automedal/agent/retry.py:with_retry. Identical posture: 3 ×
// [5,15,30]s by default.
func WithRetry[T any](ctx context.Context, fn func() (T, error), opts RetryOpts) (T, error) {
	if opts.Attempts <= 0 {
		opts.Attempts = defaultRetryAttempts
	}
	if len(opts.Delays) == 0 {
		opts.Delays = defaultRetryDelays
	}
	var zero T
	var lastErr error
	for attempt := 1; attempt <= opts.Attempts; attempt++ {
		out, err := fn()
		if err == nil {
			return out, nil
		}
		lastErr = err
		if attempt >= opts.Attempts || !IsTransientError(err) {
			return zero, err
		}
		idx := attempt - 1
		if idx >= len(opts.Delays) {
			idx = len(opts.Delays) - 1
		}
		delay := opts.Delays[idx]
		if opts.Sink != nil {
			opts.Sink.ToolLog("retry", fmt.Sprintf(
				"%s: transient error (attempt %d/%d) — retrying in %s: %s",
				opts.Label, attempt, opts.Attempts, delay, err,
			))
		}
		select {
		case <-ctx.Done():
			return zero, ctx.Err()
		case <-time.After(delay):
		}
	}
	return zero, lastErr
}
